// @citation-js/plugin-wikidata/lib/entity.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseEntity = parseEntity;
exports.parseAsync = exports.parseEntitiesAsync = parseEntitiesAsync;
exports.default = exports.parse = exports.parseEntities = parseEntities;

var _core = require("@citation-js/core");

var response = _interopRequireWildcard(require("./response"));

var _prop = require("./prop");

var _props = require("./props");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function resolveProp(prop_, entity, unkown) {
  function resolve([prop, ...parts], {
    claims
  }) {
    if (!parts.length) {
      return claims[prop];
    } else if (claims[prop] && claims[prop].length) {
      return resolve(parts, claims[prop][0].value);
    }
  }

  const parts = prop_.split('.');
  unkown.delete(parts[0]);
  return resolve(parts, entity);
}

function prepareValue(statement, entity, unkown) {
  if (typeof statement !== 'object') {
    const value = resolveProp(statement, entity, unkown);
    return value && value[0].value;
  }

  const values = [].concat(...statement.props.map(prop => resolveProp(prop, entity, unkown)).filter(Boolean));

  if (statement.values === 'all') {
    return values[0] && values;
  } else {
    return values[0] && values[0].value;
  }
}

function parseEntity(entity) {
  const data = {
    id: entity.id,
    _wikiId: entity.id,
    source: 'Wikidata'
  };
  const unkown = new Set(Object.keys(entity.claims));

  for (let prop in _props.props) {
    const input = prepareValue(_props.props[prop], entity, unkown);

    if (input) {
      const output = (0, _prop.parseProp)(prop, input, entity);

      if (output) {
        data[prop] = output;
      }
    }
  }

  for (let prop of unkown) {
    if (prop in _props.ignoredProps) {
      continue;
    }

    _core.logger.unmapped('[plugin-wikidata]', 'property', prop);
  }

  if (!data.title) {
    data.title = (0, _prop.getLabel)(entity);
  }

  if (data['reviewed-title'] || data['reviewed-author']) {
    if (data.type.slice(0, 6) !== 'review') {
      data.type = 'review';
    }

    delete data.keyword;
  }

  if (data.recipient) {
    data.type = 'personal_communication';
  }

  if (data.event) {
    data.type = 'paper-conference';
  }

  return data;
}

async function parseEntitiesAsync({
  entities
}) {
  return (await response.parseAsync(entities)).map(parseEntity);
}

function parseEntities({
  entities
}) {
  return response.parse(entities).map(parseEntity);
}
